### ISSUE TITLE
Doing cross-compiled bundles?

### ISSUE THREAD

**Issue by umbertov** [NixOS/bundlers#8]
Is it possible to bundle e.g. a `.deb` package for `aarch64` system when compiling from an x86 system? 

I'm testing with a very basic cmake c++ hello world, but it seems to me that, even if `nix-build` or `nix build -f default.nix` do produce `aarch64` binaries (i checked with the `file` command), if i try doing `nix bundle -f default.nix --bundler github:NixOS/bundlers#toDEB`, i end up with a `.deb` file for `amd64`.

**Comment by tennox** [comment]
You checkout https://nixos.wiki/wiki/Cross_Compiling - probably something like:
```
nix-build '<nixpkgs>' -A pkgsCross.aarch64-multiplatform.curl
```
if you want to build curl

I think it's unrelated to bundling

**Comment by doronbehar** [comment]
> I end up with a `.deb` file for `amd64`.

I too experience this issue!

I bundled a cross compiled package from a flake:

```
$ nix build -L .\?submodules=1\#pythonEnv-armv7l-hf-multiplatform
$ file result/bin/python
result/bin/python: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /nix/store/s9alhnp7zb4pbjlp1ap730drmr3w50lb-glibc-armv7l-unknown-linux-gnueabihf-2.35-224/lib/ld-linux-armhf.so.3, for GNU/Linux 3.10.0, not stripped
```

I build the bundle with:

```
$ nix bundle --bundler bundlers#toDEB .\?submodules=1\#pythonEnv-armv7l-hf-multiplatform
```

And when I `try to install the `.deb` file on the target machine, I get:

```
# dpkg -i *deb
dpkg: error processing archive python3-armv7l-unknown-linux-gnueabihf_1.0_amd64.deb (--install):
 package architecture (amd64) does not match system (armhf)
```

Also the `_amd64.deb` suffix indicates this error in detecting the target architecture of the `.deb` file.

**Comment by hacker1024** [comment]
Yep, the issue is that the `.deb` file targets amd64 in its metadata, even if the derivation has arm64 binaries.

**Comment by doronbehar** [comment]
I managed to generate a `.deb` file for cross compiled binaries without [`fpm`](https://github.com/jordansissel/fpm), which NixOS/bundlers [uses via it's `nix-utils` flake dependency](https://github.com/juliosueiras-nix/nix-utils/blob/master/utils/rpm-deb/default.nix).

Here's a link to my implementation:

https://github.com/doronbehar/scipy/blob/7b55b0189f4ee22aba365be3ceac55fa8c3e50b3/flake.nix#L147-L203

Note that in order to use it you have to know your target architecture's name in Debian's terminology - there is not a ready to use Nix table for that in Nixpkgs' `lib.systems` or something like that. I cross compiled my package for `armv7l-hf-multiplatform` (Nix terminology) - `armhf` (Debian terminology).

**Comment by Eveeifyeve** [comment]
I was discussing this with @tomberek a long time ago, the idea is chain pkgsCross in bundlers like the example: `nix bundle -B github:NixOS/bundlers#pkgsCross -B github:NixOS/bundlers#toAppImage nixpkgs#hello`

**NOTE: `-B` is a new addition to nix 2.34**

**Comment by doronbehar** [comment]
@Eveeifyeve thanks for commenting. 1 Question: I couldn't find the `-B` documentation in `nix bundle --help` manpage. I did find the `--bundler` option which seems like the equivalent. I don't understand though, what does it mean to use it twice (as in your example with `pkgsCross` and `toAppImage` from `github:NixOS/bundlers`.

**Comment by Eveeifyeve** [comment]
> [@Eveeifyeve](https://github.com/Eveeifyeve) thanks for commenting. 1 Question: I couldn't find the `-B` documentation in `nix bundle --help` manpage. I did find the `--bundler` option which seems like the equivalent. 

What version of nix do have installed? This option was only introduced in 2.34.

> I don't understand though, what does it mean to use it twice (as in your example with `pkgsCross` and `toAppImage` from `github:NixOS/bundlers`.

So first it means it will first cross compile the derivation using the cross bundler then it will feed that output and turn that cross compiled derviation to a appimage.

**Comment by doronbehar** [comment]
> What version of nix do have installed? This option was only introduced in 2.34.

OK I see. I don't see it documented there though:

```
nix run -Lf. nixVersions.nix_2_34 -- bundle --help | grep -- '-B'
```

Prints nothing. I think opening an issue in Nix would be nice?

**Comment by doronbehar** [comment]
Hmm it seems according to [here](https://nix.dev/manual/nix/development/command-ref/new-cli/nix3-bundle.html#options), that `nix bundle --bundler` is the same as `nix bundle -B`. I won't open a Nix issue for this - it looks fixed in the current Nix source.

Back to you suggestion @Eveeifyeve it seems the functionality you demonstrate shouldn't depend on Nix 2.34 and should be available with any Nix that supports `bundle --bundler` argument.
